File: iPlant.FMS.Service/DAO/OMS/OMSCommandDAO.java
package com.mes.aps.server.serviceimpl.dao.oms;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
import org.springframework.jdbc.core.namedparam.SqlParameterSource;
import org.springframework.jdbc.support.GeneratedKeyHolder;
import org.springframework.jdbc.support.KeyHolder;

import com.mes.aps.server.service.mesenum.MESDBSource;
import com.mes.aps.server.service.mesenum.MESException;
import com.mes.aps.server.service.po.OutResult;
import com.mes.aps.server.service.po.ServiceResult;
import com.mes.aps.server.service.po.bms.BMSEmployee;
import com.mes.aps.server.service.po.oms.OMSCommand;
import com.mes.aps.server.service.utils.StringUtils;
import com.mes.aps.server.serviceimpl.dao.BaseDAO;
import com.mes.aps.server.serviceimpl.dao.oms.OMSCommandDAO;
import com.mes.aps.server.serviceimpl.utils.aps.APSConstants;

public class OMSCommandDAO extends BaseDAO {
	private static Logger logger = LoggerFactory.getLogger(OMSCommandDAO.class);

	private static OMSCommandDAO Instance = null;

	public static OMSCommandDAO getInstance() {
		if (Instance == null)
			Instance = new OMSCommandDAO();
		return Instance;
	}

	public int Update(BMSEmployee wLoginUser, OMSCommand wOMSCommand, OutResult<Integer> wErrorCode) {
		int wResult = 0;
		try {
			ServiceResult<String> wInstance = GetDataBaseName(wLoginUser.getCompanyID(), MESDBSource.APS,
					wLoginUser.getID(), 0);
			wErrorCode.set(wInstance.ErrorCode);
			if (wErrorCode.Result != 0) {
				return wResult;
			}

			if (wOMSCommand == null) {
				return 0;
			}
			String wSQL = "";
			if (wOMSCommand.getID() <= 0) {

				wOMSCommand.No = StringUtils.Format("CM-{0}-{1}",
						String.format("%04d", this.OMS_GetCommandMaxNo(wLoginUser, wErrorCode) + 1),
						StringUtils.parseCalendarToString(Calendar.getInstance(), "MM-yyyy"));

				wSQL = StringUtils.Format(
						"INSERT INTO {0}.oms_command(CustomerID,ContactCode,No,Status,LinkManID,EditorID,EditTime,"
								+ "Active,AuditorID,AuditTime,CreatorID,CreateTime,FactoryID,BusinessUnitID,FQTY,WBSNo,WorkShopID)"
								+ " VALUES(:CustomerID,:ContactCode,:No,:Status,:LinkManID,:EditorID,:EditTime,:Active,:AuditorID,"
								+ ":AuditTime,:CreatorID,:CreateTime,:FactoryID,:BusinessUnitID,:FQTY,:WBSNo,:WorkShopID);",

						wInstance.Result);
			} else {
				wSQL = StringUtils.Format(
						"UPDATE {0}.oms_command SET CustomerID = :CustomerID,ContactCode = :ContactCode,No = :No,WorkShopID=:WorkShopID,"
								+ "Status = :Status,LinkManID = :LinkManID,EditorID = :EditorID,EditTime = :EditTime,Active = :Active,"
								+ "AuditorID = :AuditorID,AuditTime = :AuditTime,CreatorID = :CreatorID,CreateTime = :CreateTime,FactoryID = :FactoryID,"
								+ "BusinessUnitID = :BusinessUnitID,FQTY = :FQTY,WBSNo=:WBSNo WHERE ID = :ID;",
						new Object[] {

								wInstance.Result });
			}
			wSQL = DMLChange(wSQL);

			Map<String, Object> wParamMap = new HashMap<>();

			wParamMap.put("ID", wOMSCommand.ID);
			wParamMap.put("WorkShopID", wOMSCommand.WorkShopID);
			wParamMap.put("CustomerID", wOMSCommand.CustomerID);
			wParamMap.put("ContactCode", wOMSCommand.ContactCode);
			wParamMap.put("No", wOMSCommand.No);
			wParamMap.put("Status", wOMSCommand.Status);
			wParamMap.put("LinkManID", wOMSCommand.LinkManID);
			wParamMap.put("EditorID", wOMSCommand.EditorID);
			wParamMap.put("EditTime", wOMSCommand.EditTime);
			wParamMap.put("Active", wOMSCommand.Active);
			wParamMap.put("AuditorID", wOMSCommand.AuditorID);
			wParamMap.put("AuditTime", wOMSCommand.AuditTime);
			wParamMap.put("CreatorID", wOMSCommand.CreatorID);
			wParamMap.put("CreateTime", wOMSCommand.CreateTime);
			wParamMap.put("FactoryID", wOMSCommand.FactoryID);
			wParamMap.put("BusinessUnitID", wOMSCommand.BusinessUnitID);
			wParamMap.put("FQTY", wOMSCommand.FQTY);
			wParamMap.put("WBSNo", wOMSCommand.WBSNo);

			GeneratedKeyHolder generatedKeyHolder = new GeneratedKeyHolder();
			MapSqlParameterSource mapSqlParameterSource = new MapSqlParameterSource(wParamMap);

			this.nameJdbcTemplate.update(wSQL, (SqlParameterSource) mapSqlParameterSource,
					(KeyHolder) generatedKeyHolder);

			if (wOMSCommand.getID() <= 0) {
				wResult = generatedKeyHolder.getKey().intValue();
				wOMSCommand.setID(wResult);
			} else {
				wResult = wOMSCommand.getID();
			}
		} catch (Exception ex) {
			wErrorCode.set(MESException.DBSQL.getValue());
			logger.Error(System.Reflection.MethodBase.GetCurrentMethod().Name, ex);
		}
		return wResult;
	}

	private int OMS_GetCommandMaxNo(BMSEmployee wLoginUser, OutResult<Integer> wErrorCode) {
		int wResult = 0;
		try {

			ServiceResult<String> wInstance = GetDataBaseName(wLoginUser.getCompanyID(), MESDBSource.APS,
					wLoginUser.getID(), 0);

			wErrorCode.set(wInstance.ErrorCode);
			if (wErrorCode.Result != 0) {
				return wResult;
			}

			Calendar wStartTime = Calendar.getInstance();
			wStartTime.set(Calendar.DAY_OF_MONTH, 1);
			wStartTime.set(Calendar.HOUR_OF_DAY, 0);
			wStartTime.set(Calendar.MINUTE, 0);
			wStartTime.set(Calendar.SECOND, 0);

			Calendar wEndTime = (Calendar) wStartTime.clone();
			wEndTime.add(Calendar.MONTH, 1);

			Map<String, Object> wParamMap = new HashMap<>();

			wParamMap.put("StartTime", wStartTime);
			wParamMap.put("EndTime", wEndTime);
			String wSql = StringUtils.Format(
					"select ID,No from {0}.oms_command WHERE CreateTime between :StartTime and :EndTime ;",
					wInstance.Result);

			List<Map<String, Object>> wQueryResult = this.nameJdbcTemplate.queryForList(wSql, wParamMap);
			String wNo = "";
			int wNum = 0;
			String wEndWiths = StringUtils.parseCalendarToString(wStartTime, "MM-yyyy");
			for (Map<String, Object> wReader : wQueryResult) {

				wNo = StringUtils.parseString(wReader["No"]);
				if (!wNo.endsWith(wEndWiths))
					continue;

				wNo = wNo.substring(3, 7);
				wNum = StringUtils.parseInt(wNo);
				if (wResult < wNum)
					wResult = wNum;
			}
		} catch (Exception ex) {
			wErrorCode.set(MESException.DBSQL.getValue());
			logger.Error(System.Reflection.MethodBase.GetCurrentMethod().Name, ex);
		}
		return wResult;
	}

	public ServiceResult<Integer> DeleteList(BMSEmployee wLoginUser, List<OMSCommand> wList,
			OutResult<Integer> wErrorCode) {
		ServiceResult<Integer> wResult = new ServiceResult<>(0);
		try {
			ServiceResult<String> wInstance = GetDataBaseName(wLoginUser.getCompanyID(), MESDBSource.APS,
					wLoginUser.getID(), 0);
			wErrorCode.set(wInstance.ErrorCode);
			if (wErrorCode.Result != 0) {
				return wResult;
			}

			if (wList == null || wList.size() <= 0) {
				return wResult;
			}
			List<String> wIDList = new ArrayList<>();
			for (OMSCommand wItem : wList) {
				wIDList.add(String.valueOf(wItem.ID));
			}
			String wSql = StringUtils.Format("delete from {1}.oms_command WHERE ID IN({0}) ;",
					String.join(",", wIDList), wInstance.Result);
			ExecuteSqlTransaction(wSql);
		} catch (Exception ex) {
			wErrorCode.set(MESException.DBSQL.getValue());
			logger.Error(System.Reflection.MethodBase.GetCurrentMethod().Name, ex);
		}
		return wResult;
	}

	public OMSCommand SelectByID(BMSEmployee wLoginUser, int wID, OutResult<Integer> wErrorCode) {
		OMSCommand wResult = new OMSCommand();
		try {
			ServiceResult<String> wInstance = GetDataBaseName(wLoginUser.getCompanyID(), MESDBSource.APS,
					wLoginUser.getID(), 0);
			wErrorCode.set(wInstance.ErrorCode);
			if (wErrorCode.Result != 0) {
				return wResult;
			}

			List<OMSCommand> wList = SelectList(wLoginUser, wID, "", "", -1, -1, -1, -1, -1, null, null, wErrorCode);
			if (wList == null || wList.size() != 1)
				return wResult;
			wResult = wList[0];
		} catch (Exception e) {
			wErrorCode.set(MESException.DBSQL.getValue());
			logger.Error(System.Reflection.MethodBase.GetCurrentMethod().Name, e);
		}
		return wResult;
	}

	public OMSCommand SelectByID(BMSEmployee wLoginUser, String wNo, OutResult<Integer> wErrorCode) {
		OMSCommand wResult = new OMSCommand();
		try {
			ServiceResult<String> wInstance = GetDataBaseName(wLoginUser.getCompanyID(), MESDBSource.APS,
					wLoginUser.getID(), 0);
			wErrorCode.set(wInstance.ErrorCode);
			if (wErrorCode.Result != 0) {
				return wResult;
			}

			List<OMSCommand> wList = SelectList(wLoginUser, -1, wNo, "", -1, -1, -1, -1, -1, null, null, wErrorCode);
			if (wList == null || wList.size() != 1)
				return wResult;
			wResult = wList[0];
		} catch (Exception e) {
			wErrorCode.set(MESException.DBSQL.getValue());
			logger.Error(System.Reflection.MethodBase.GetCurrentMethod().Name, e);
		}
		return wResult;
	}

	public List<OMSCommand> SelectList(BMSEmployee wLoginUser, int wID, String wNo, String wWBSNo, int wFactoryID,
			int wBusinessUnitID, int wWorkShopID, int wCustomerID, int wActive, Calendar wStartTime, Calendar wEndTime,
			OutResult<Integer> wErrorCode) {
		List<OMSCommand> wResultList = new ArrayList<>();
		try {
			ServiceResult<String> wInstance = GetDataBaseName(wLoginUser.getCompanyID(), MESDBSource.APS,
					wLoginUser.getID(), 0);
			wErrorCode.set(wInstance.ErrorCode);
			if (wErrorCode.Result != 0) {
				return wResultList;
			}

			Calendar wBaseTime = Calendar.getInstance();
			wBaseTime.set(2000, 1, 1);
			if ( wStartTime.compareTo(wBaseTime) < 0)
				wStartTime = wBaseTime;
			if ( wEndTime.compareTo(wBaseTime) < 0)
				wEndTime = wBaseTime;
			if (wStartTime.compareTo(wEndTime) > 0) {
				return wResultList;
			}

			String wSQL = StringUtils.Format(
					"SELECT t.*,Count(t1.ID) as OrderCount,Sum(t1.FQTYPlan) as FQTYPlan,Sum(t1.FQTYActual) as FQTYActual FROM {0}.oms_command t"
							+ " Left join {0}.oms_order t1 on t1.CommandID=t.ID WHERE  1=1  "
							+ " and ( :wID <= 0 or :wID = t.ID )  and ( :wNo <= 0 or :wNo = t.No ) "
							+ " and ( :wFactoryID <= 0 or :wFactoryID = t.FactoryID )  and ( :wWBSNo = '''' or :wWBSNo = t.WBSNo )"
							+ " and ( :wWorkShopID <= 0 or :wWorkShopID = t.WorkShopID ) "
							+ " and ( :wCustomerID <= 0 or :wCustomerID = t.CustomerID ) "
							+ " and ( :wBusinessUnitID <= 0 or :wBusinessUnitID = t.BusinessUnitID ) "
							+ " and ( :wStartTime <= str_to_date(''2010-01-01'', ''%Y-%m-%d'') or :wStartTime <= t.EditTime) "
							+ " and ( :wEndTime <= str_to_date(''2010-01-01'', ''%Y-%m-%d'') or :wEndTime >= t.CreateTime) group by t.ID;",
					wInstance.Result);
			Map<String, Object> wParamMap = new HashMap<>();

			wParamMap.put("wID", wID);
			wParamMap.put("wNo", wNo);
			wParamMap.put("wActive", wActive);
			wParamMap.put("wFactoryID", wFactoryID);
			wParamMap.put("wCustomerID", wCustomerID);
			wParamMap.put("wWBSNo", wWBSNo);
			wParamMap.put("wWorkShopID", wWorkShopID);
			wParamMap.put("wBusinessUnitID", wBusinessUnitID);
			wParamMap.put("wStartTime", wStartTime);
			wParamMap.put("wEndTime", wEndTime);

			wSQL = DMLChange(wSQL);

			List<Map<String, Object>> wQueryResult = this.nameJdbcTemplate.queryForList(wSQL, wParamMap);

			for (Map<String, Object> wReader : wQueryResult) {
				OMSCommand wItem = new OMSCommand();

				wItem.ID = StringUtils.parseInt(wReader["ID"]);
				wItem.CustomerID = StringUtils.parseInt(wReader["CustomerID"]);
				wItem.ContactCode = StringUtils.parseString(wReader["ContactCode"]);
				wItem.No = StringUtils.parseString(wReader["No"]);
				wItem.Status = StringUtils.parseInt(wReader["Status"]);
				wItem.LinkManID = StringUtils.parseInt(wReader["LinkManID"]);
				wItem.WorkShopID = StringUtils.parseInt(wReader["WorkShopID"]);
				wItem.WorkShopName = APSConstants.GetFMCWorkShopName(wItem.WorkShopID);
				wItem.EditorID = StringUtils.parseInt(wReader["EditorID"]);
				wItem.EditTime = StringUtils.parseCalendar(wReader["EditTime"]);
				wItem.Active = StringUtils.parseInt(wReader["Active"]);
				wItem.AuditorID = StringUtils.parseInt(wReader["AuditorID"]);
				wItem.AuditTime = StringUtils.parseCalendar(wReader["AuditTime"]);
				wItem.CreatorID = StringUtils.parseInt(wReader["CreatorID"]);
				wItem.CreateTime = StringUtils.parseCalendar(wReader["CreateTime"]);
				wItem.FactoryID = StringUtils.parseInt(wReader["FactoryID"]);
				wItem.BusinessUnitID = StringUtils.parseInt(wReader["BusinessUnitID"]);
				wItem.FQTY = StringUtils.parseDouble(wReader["FQTY"]);
				wItem.OrderCount = StringUtils.parseInt(wReader["OrderCount"]);
				wItem.FQTYPlan = StringUtils.parseDouble(wReader["FQTYPlan"]);
				wItem.FQTYActual = StringUtils.parseDouble(wReader["FQTYActual"]);
				wItem.WBSNo = StringUtils.parseString(wReader["WBSNo"]);
				wItem.Customer = APSConstants.GetCRMCustomerName(wItem.CustomerID);
				wResultList.add(wItem);
			}
		} catch (Exception ex) {
			wErrorCode.set(MESException.DBSQL.getValue());
			logger.Error(System.Reflection.MethodBase.GetCurrentMethod().Name, ex);
		}
		return wResultList;
	}

	public ServiceResult<Integer> Active(BMSEmployee wLoginUser, List<Integer> wIDList, int wActive,
			OutResult<Integer> wErrorCode) {
		ServiceResult<Integer> wResult = new ServiceResult<>(0);
		try {
			ServiceResult<String> wInstance = this.GetDataBaseName(wLoginUser, MESDBSource.APS);
			wErrorCode.set(wInstance.ErrorCode);
			if (wErrorCode.Result != 0) {
				return wResult;
			}

			if (wIDList == null || wIDList.size() <= 0)
				return wResult;
			for (Integer wItem : wIDList) {
				OMSCommand wOMSCommand = SelectByID(wLoginUser, wItem, wErrorCode);
				if (wOMSCommand == null || wOMSCommand.ID <= 0) {
					continue;
				}
				if (wActive == 2 && wOMSCommand.Active != 1) {
					wErrorCode.set(MESException.Logic.getValue());
					return wResult;
				}
				wOMSCommand.Active = wActive;
				long wID = Update(wLoginUser, wOMSCommand, wErrorCode);
				if (wID <= 0L)
					break;
			}
		} catch (Exception e) {
			wErrorCode.set(MESException.DBSQL.getValue());
			logger.Error(System.Reflection.MethodBase.GetCurrentMethod().Name, e);
		}
		return wResult;
	}

}

File: iPlant.FMS.Service/DAO/OMS/OMSCommandDAO.cs
using iPlant.Common.Tools;
using iPlant.FMS.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace iPlant.SCADA.Service
{
    public class OMSCommandDAO : BaseDAO
    {

        private static readonly log4net.ILog logger = log4net.LogManager.GetLogger(typeof(OMSCommandDAO));
        private static OMSCommandDAO Instance = null;

        private OMSCommandDAO() : base()
        {

        }

        public static OMSCommandDAO getInstance()
        {
            if (Instance == null)
                Instance = new OMSCommandDAO();
            return Instance;
        }
        // 工厂&事业部&车间&产线&工位
        // 工厂
        private OMSCommand OMS_CheckCommand(BMSEmployee wLoginUser, OMSCommand wCommand,
                OutResult<Int32> wErrorCode)
        {
            OMSCommand wCommandDB = new OMSCommand();
            wErrorCode.set(0);

            try
            {
                wErrorCode.set(0); 	 String wInstance =iPlant.Data.EF.MESDBSource.Basic.getDBName();

                if (wErrorCode.Result == 0)
                {
                    // Step0:查询
                    Dictionary<String, Object> wParms = new Dictionary<String, Object>();
                    String wSQLText = StringUtils.Format("Select ID from {0}.oms_command ", wInstance)
                                + " where ID!=@ID and (WBSNo=@WBSNo or PartNo = @PartNo )";
                    wParms.Clear();
                    wParms.Add("ID", wCommand.ID);
                    wParms.Add("PartNo", wCommand.PartNo);
                    wParms.Add("WBSNo", wCommand.WBSNo);

                    wSQLText = this.DMLChange(wSQLText);
                    List<Dictionary<String, Object>> wQueryResultList = mDBPool.queryForList(wSQLText, wParms);
                    foreach (Dictionary<String, Object> wSqlDataReader in wQueryResultList)
                    {
                        wCommandDB.ID = StringUtils.parseInt(wSqlDataReader["ID"]);

                    }

                }
            }
            catch (Exception ex)
            {
                wErrorCode.set(MESException.DBSQL.Value);
                logger.Error("OMS_CheckCommandByCode", ex);

            }
            return wCommandDB;
        }
        public void OMS_UpdateCommand(BMSEmployee wLoginUser, OMSCommand wOMSCommand, OutResult<Int32> wErrorCode)
        {
            try
            {


                if (wOMSCommand == null || StringUtils.isEmpty(wOMSCommand.PartNo) || StringUtils.isEmpty(wOMSCommand.WBSNo))
                {
                    wErrorCode.Result = MESException.Parameter.getValue();
                    return;
                }

                OMSCommand wOMSCommandDB = this.OMS_CheckCommand(wLoginUser, wOMSCommand, wErrorCode);
                if (wOMSCommandDB != null && wOMSCommandDB.ID > 0)
                {
                    wErrorCode.Result = MESException.Duplication.getValue();
                    return;
                }

                wErrorCode.set(0); 	 String wInstance =iPlant.Data.EF.MESDBSource.APS.getDBName();
                if (wErrorCode.Result != 0)
                {
                    return;
                }


                Dictionary<String, Object> wParamMap = new Dictionary<String, Object>();


                wParamMap.Add("WBSNo", wOMSCommand.WBSNo);
                wParamMap.Add("PartNo", wOMSCommand.PartNo);
                wParamMap.Add("ProductID", wOMSCommand.ProductID);
                wParamMap.Add("CustomerID", wOMSCommand.CustomerID);
                wParamMap.Add("ContactCode", wOMSCommand.ContactCode);
                wParamMap.Add("LinkManID", wOMSCommand.LinkManID);
                wParamMap.Add("EditorID", wLoginUser.ID);
                wParamMap.Add("EditTime", DateTime.Now);
                wParamMap.Add("FactoryID", wOMSCommand.FactoryID);
                wParamMap.Add("BusinessUnitID", wOMSCommand.BusinessUnitID);
                wParamMap.Add("WorkShopID", wOMSCommand.WorkShopID);

                if (wOMSCommand.ID <= 0)
                {

                    wParamMap.Add("CreatorID", wLoginUser.ID);
                    wParamMap.Add("CreateTime", DateTime.Now);
                    wOMSCommand.ID = (int)this.Insert(StringUtils.Format(
                            "{0}.oms_command", wInstance), wParamMap);
                }
                else
                {
                    wParamMap.Add("ID", wOMSCommand.ID);

                    this.Update(StringUtils.Format(
                          "{0}.oms_command", wInstance), "ID", wParamMap);
                }

            }
            catch (Exception ex)
            {
                wErrorCode.set(MESException.DBSQL.getValue());
                logger.Error(System.Reflection.MethodBase.GetCurrentMethod().Name, ex);
            }
        }

        public ServiceResult<Int32> OMS_DeleteCommandList(BMSEmployee wLoginUser, List<OMSCommand> wList,
                OutResult<Int32> wErrorCode)
        {
            ServiceResult<Int32> wResult = new ServiceResult<Int32>(0);
            try
            {
                wErrorCode.set(0); 	 String wInstance =iPlant.Data.EF.MESDBSource.APS.getDBName();
                if (wErrorCode.Result != 0)
                {
                    return wResult;
                }

                if (wList == null || wList.Count <= 0)
                {
                    return wResult;
                }
                List<Int32> wIDList = new List<Int32>();
                foreach (OMSCommand wItem in wList)
                {
                    if (wItem == null || wItem.ID <= 0 || wItem.OrderCount > 0)
                        continue;
                    wIDList.Add(wItem.ID);
                }
                if (wIDList.Count <= 0)
                    return wResult;
                String wSql = StringUtils.Format("delete from {1}.oms_command WHERE ID IN({0}) ;",
                        StringUtils.Join(",", wIDList), wInstance);
                ExecuteSqlTransaction(wSql);
            }
            catch (Exception ex)
            {
                wErrorCode.set(MESException.DBSQL.getValue());
                logger.Error(System.Reflection.MethodBase.GetCurrentMethod().Name, ex);
            }
            return wResult;
        }

        public OMSCommand OMS_SelectCommandByID(BMSEmployee wLoginUser, int wID, OutResult<Int32> wErrorCode)
        {
            OMSCommand wResult = new OMSCommand();
            try
            {
                wErrorCode.set(0); 	 String wInstance =iPlant.Data.EF.MESDBSource.APS.getDBName();
                if (wErrorCode.Result != 0)
                {
                    return wResult;
                }
                DateTime wBaseTime = new DateTime(2000, 1, 1);
                List<OMSCommand> wList = OMS_SelectCommandList(wLoginUser, wID, "", "", -1, -1, -1, -1, -1, wBaseTime, wBaseTime, wErrorCode);
                if (wList == null || wList.Count != 1)
                    return wResult;
                wResult = wList[0];
            }
            catch (Exception e)
            {
                wErrorCode.set(MESException.DBSQL.getValue());
                logger.Error(System.Reflection.MethodBase.GetCurrentMethod().Name, e);
            }
            return wResult;
        }

        public OMSCommand OMS_SelectCommandByCode(BMSEmployee wLoginUser, String wWBSNo, OutResult<Int32> wErrorCode)
        {
            OMSCommand wResult = new OMSCommand();
            try
            {
                wErrorCode.set(0); 	 String wInstance =iPlant.Data.EF.MESDBSource.APS.getDBName();
                if (wErrorCode.Result != 0)
                {
                    return wResult;
                }
                DateTime wBaseTime = new DateTime(2000, 1, 1);

                List<OMSCommand> wList = OMS_SelectCommandList(wLoginUser, -1, "", wWBSNo, -1, -1, -1, -1, -1, wBaseTime, wBaseTime, wErrorCode);
                if (wList == null || wList.Count != 1)
                    return wResult;
                wResult = wList[0];
            }
            catch (Exception e)
            {
                wErrorCode.set(MESException.DBSQL.getValue());
                logger.Error(System.Reflection.MethodBase.GetCurrentMethod().Name, e);
            }
            return wResult;
        }

        public OMSCommand OMS_SelectCommandByPartNo(BMSEmployee wLoginUser, String PartNo, OutResult<Int32> wErrorCode)
        {
            OMSCommand wResult = new OMSCommand();
            try
            {
                wErrorCode.set(0); 	 String wInstance =iPlant.Data.EF.MESDBSource.APS.getDBName();
                if (wErrorCode.Result != 0)
                {
                    return wResult;
                }
                DateTime wBaseTime = new DateTime(2000, 1, 1);

                List<OMSCommand> wList = OMS_SelectCommandList(wLoginUser, -1, PartNo, "", -1, -1, -1, -1, -1, wBaseTime, wBaseTime, wErrorCode);
                if (wList == null || wList.Count != 1)
                    return wResult;
                wResult = wList[0];
            }
            catch (Exception e)
            {
                wErrorCode.set(MESException.DBSQL.getValue());
                logger.Error(System.Reflection.MethodBase.GetCurrentMethod().Name, e);
            }
            return wResult;
        }

        private List<OMSCommand> OMS_SelectCommandList(BMSEmployee wLoginUser, int wID, String wPartNo, String wWBSNo, int wFactoryID,
                int wBusinessUnitID, int wWorkShopID, int wCustomerID, int wProductID, DateTime wStartTime, DateTime wEndTime,
                OutResult<Int32> wErrorCode)
        {
            List<OMSCommand> wResultList = new List<OMSCommand>();
            try
            {
                wErrorCode.set(0); 	 String wInstance =iPlant.Data.EF.MESDBSource.APS.getDBName();
                if (wErrorCode.Result != 0)
                {
                    return wResultList;
                }

                DateTime wBaseTime = new DateTime(2000, 1, 1);

                if ( wStartTime.CompareTo(wBaseTime) < 0)
                    wStartTime = wBaseTime;
                if ( wEndTime.CompareTo(wBaseTime) < 0)
                    wEndTime = wBaseTime;
                if (wStartTime.CompareTo(wEndTime) > 0)
                {
                    return wResultList;
                }

                String wSQL = StringUtils.Format(
                        "SELECT t.*,Count(t1.ID) as OrderCount,t2.ProductNo,t3.CustomerName ,"
                                + " t3.CustomerCode ,t4.Name as  Factory,t5.Name as WorkShopName,"
                                + " t6.Name as Creator,t7.Name as Editor FROM {0}.oms_command t"
                                + " Left join {0}.oms_order t1 on t1.CommandID=t.ID "
                                + " Left join {0}.fpc_product t2 on t.ProductID=t2.ID "
                                + " Left join {0}.crm_customer t3 on t.CustomerID=t3.ID "
                                + " Left join {0}.fmc_factory t4 on t.FactoryID=t4.ID "
                                + " Left join {0}.fmc_workshop t5 on t.WorkShopID=t5.ID "
                                + " Left join {0}.mbs_user t6 on t.CreatorID=t6.ID "
                                + " Left join {0}.mbs_user t7 on t.EditorID=t7.ID WHERE  1=1  "
                                + " and ( @wID <= 0 or @wID = t.ID )  and ( @wWBSNo = '' or @wWBSNo = t.WBSNo ) "
                                + " and ( @wPartNo = '' or @wPartNo = t.PartNo )"
                                + " and ( @wFactoryID <= 0 or @wFactoryID = t.FactoryID )   "
                                + " and ( @wProductID <= 0 or @wProductID = t.ProductID )   "
                                + " and ( @wWorkShopID <= 0 or @wWorkShopID = t.WorkShopID ) "
                                + " and ( @wCustomerID <= 0 or @wCustomerID = t.CustomerID ) "
                                + " and ( @wBusinessUnitID <= 0 or @wBusinessUnitID = t.BusinessUnitID ) "
                                + " and ( @wStartTime <= str_to_date('2010-01-01', '%Y-%m-%d') or @wStartTime <= t.EditTime) "
                                + " and ( @wEndTime <= str_to_date('2010-01-01', '%Y-%m-%d') or @wEndTime >= t.CreateTime) group by t.ID;",
                        wInstance);
                Dictionary<String, Object> wParamMap = new Dictionary<String, Object>();

                wParamMap.Add("wID", wID);
                wParamMap.Add("wWBSNo", wWBSNo);
                wParamMap.Add("wPartNo", wPartNo);
                wParamMap.Add("wFactoryID", wFactoryID);
                wParamMap.Add("wProductID", wProductID);
                wParamMap.Add("wCustomerID", wCustomerID);
                wParamMap.Add("wWorkShopID", wWorkShopID);
                wParamMap.Add("wBusinessUnitID", wBusinessUnitID);
                wParamMap.Add("wStartTime", wStartTime);
                wParamMap.Add("wEndTime", wEndTime);

                wSQL = DMLChange(wSQL);

                List<Dictionary<String, Object>> wQueryResult = this.mDBPool.queryForList(wSQL, wParamMap);

                foreach (Dictionary<String, Object> wReader in wQueryResult)
                {
                    OMSCommand wItem = new OMSCommand();

                    wItem.ID = StringUtils.parseInt(wReader["ID"]);
                    wItem.CustomerID = StringUtils.parseInt(wReader["CustomerID"]);
                    wItem.CustomerName = StringUtils.parseString(wReader["CustomerName"]);
                    wItem.CustomerCode = StringUtils.parseString(wReader["CustomerCode"]);
                    wItem.ContactCode = StringUtils.parseString(wReader["ContactCode"]);
                    wItem.PartNo = StringUtils.parseString(wReader["PartNo"]);
                    wItem.ProductID = StringUtils.parseInt(wReader["ProductID"]);
                    wItem.ProductNo = StringUtils.parseString(wReader["ProductNo"]);
                    wItem.LinkManID = StringUtils.parseInt(wReader["LinkManID"]);
                    wItem.WorkShopID = StringUtils.parseInt(wReader["WorkShopID"]);
                    wItem.WorkShopName = StringUtils.parseString(wReader["WorkShopName"]);
                    wItem.EditorID = StringUtils.parseInt(wReader["EditorID"]);
                    wItem.Editor = StringUtils.parseString(wReader["Editor"]);
                    wItem.EditTime = StringUtils.parseDate(wReader["EditTime"]);
                    wItem.CreatorID = StringUtils.parseInt(wReader["CreatorID"]);
                    wItem.Creator = StringUtils.parseString(wReader["Creator"]);
                    wItem.CreateTime = StringUtils.parseDate(wReader["CreateTime"]);
                    wItem.FactoryID = StringUtils.parseInt(wReader["FactoryID"]);
                    wItem.Factory = StringUtils.parseString(wReader["Factory"]);
                    wItem.BusinessUnitID = StringUtils.parseInt(wReader["BusinessUnitID"]);
                    wItem.OrderCount = StringUtils.parseInt(wReader["OrderCount"]);
                    wItem.WBSNo = StringUtils.parseString(wReader["WBSNo"]);
                    wResultList.Add(wItem);
                }
            }
            catch (Exception ex)
            {
                wErrorCode.set(MESException.DBSQL.getValue());
                logger.Error(System.Reflection.MethodBase.GetCurrentMethod().Name, ex);
            }
            return wResultList;
        }

        public List<OMSCommand> OMS_SelectCommandList(BMSEmployee wLoginUser, int wFactoryID,
                int wBusinessUnitID, int wWorkShopID, int wCustomerID, int wProductID, DateTime wStartTime, DateTime wEndTime,
                OutResult<Int32> wErrorCode)
        {
            return OMS_SelectCommandList(wLoginUser, -1, "", "", wFactoryID,
                 wBusinessUnitID, wWorkShopID, wCustomerID, wProductID, wStartTime, wEndTime,
               wErrorCode);
        }

    }
}
